package org.cytoscape.engnet.view.configurationDialogs;

import java.io.File;
import javax.swing.JFileChooser;

import org.cytoscape.engnet.controller.tasks.ExecuteEnGNetTask;
import org.cytoscape.engnet.controller.utils.CySwing2;
import org.cytoscape.work.Task;
import org.cytoscape.work.TaskManager;

/**
 * @license Apache License V2 <http://www.apache.org/licenses/LICENSE-2.0.html>
 * @author 
 */
public class ConfigurationDialog extends javax.swing.JDialog {
    private final TaskManager taskManager;
    private String sPath;
    private File currentFolder;
    private String dataFilePath;
    
    /**
     * Creates new form NewJDialog
     */
    public ConfigurationDialog(TaskManager taskManager) {        
        initComponents();
        setLocationRelativeTo(CySwing2.getDesktopJFrame());
        this.taskManager = taskManager;
    }

    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        label1 = new java.awt.Label();
        runEnGNetButton = new javax.swing.JButton();
        loadDatasetButton = new javax.swing.JButton();
        datasetsPanel = new javax.swing.JLayeredPane();
        explanationLabel = new javax.swing.JLabel();
        spearman = new javax.swing.JSpinner();
        averageT = new javax.swing.JSpinner();
        thb = new javax.swing.JSpinner();
        jLabel1 = new javax.swing.JLabel();
        jLabel2 = new javax.swing.JLabel();
        jLabel3 = new javax.swing.JLabel();
        jLabel4 = new javax.swing.JLabel();
        jLabel5 = new javax.swing.JLabel();
        kendall = new javax.swing.JSpinner();
        nmi = new javax.swing.JSpinner();
        jScrollPane1 = new javax.swing.JScrollPane();
        jPanel2 = new javax.swing.JPanel();
        genesFileTextField = new javax.swing.JTextField();


        setTitle("EnGNet");
        setResizable(false);
        
        loadDatasetButton.setText("Dataset file");
        loadDatasetButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                loadDatasetButtonActionPerformed(evt);
            }
        });
        
        genesFileTextField.setBackground(new java.awt.Color(238,238,238));
        genesFileTextField.setText("Dataset file...");
        genesFileTextField.setBorder(null);
        
        datasetsPanel.setLayer(loadDatasetButton, javax.swing.JLayeredPane.DEFAULT_LAYER);
        

        explanationLabel.setText("<html>Select a biological RNA-seq file to generate network with using EnGNet. You can select each input parameter.<html>");
        explanationLabel.setPreferredSize(new java.awt.Dimension(958, 48));

        jPanel2.setBorder(javax.swing.BorderFactory.createEtchedBorder());

        jLabel1.setText("NMI:");
        jLabel1.setToolTipText("Normalized Mutual Information (NMI) measures. the NMI is able to detect linear and also non-linear dependencies between genes. Recomended: 0.6  ");
        nmi.setModel(new javax.swing.SpinnerNumberModel(0.6, 0.0, 1.00, 0.1));
        
        jLabel2.setText("Kendall:");
        jLabel2.setToolTipText("Kendall coefficient. Kendall’s measure is also able to detect linear dependencies. Recomended: 0.7 ");
        kendall.setModel(new javax.swing.SpinnerNumberModel(0.7, 0.0, 1.00, 0.1));
        
        jLabel3.setText("Spearman:");
        jLabel3.setToolTipText("Spearman coefficient. The Spearman coefficient is able to detect linear dependencies between two genes unaffected by data distribution. Recomended: 0.7");
        spearman.setModel(new javax.swing.SpinnerNumberModel(0.7, 0.0, 1.00, 0.1));
        
        jLabel4.setText("Significant threshold:");
        jLabel4.setToolTipText("To prune de tree. Recomended: 0.7 ");
        averageT.setModel(new javax.swing.SpinnerNumberModel(0.7, 0.0, 1.00, 0.1));

        jLabel5.setText("Thβ:");
        jLabel5.setToolTipText("Thβ threshold. This threshold is a user set parameter, which is employed to determine the biological relevance level of the removed edges. Recomended: 3");
        thb.setModel(new javax.swing.SpinnerNumberModel(3, 0, null, 1));
        
        
        datasetsPanel.setLayer(loadDatasetButton, javax.swing.JLayeredPane.DEFAULT_LAYER);
        datasetsPanel.setLayer(jScrollPane1, javax.swing.JLayeredPane.DEFAULT_LAYER);
        
       
        
        runEnGNetButton.setFont(new java.awt.Font("Lucida Grande", 0, 18)); // NOI18N
        runEnGNetButton.setIcon(new javax.swing.ImageIcon(getClass().getResource("/images/play.png"))); // NOI18N
        runEnGNetButton.setText("Execute EnGNet");
        runEnGNetButton.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                runEnGNetButtonActionPerformed(evt);
            }
        });
        
        
        javax.swing.GroupLayout jPanel2Layout = new javax.swing.GroupLayout(jPanel2);
        jPanel2.setLayout(jPanel2Layout); // determina la disposicion del contenedor
        jPanel2Layout.setHorizontalGroup(
            jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING) // each row of components should be justified 
            .addGroup(jPanel2Layout.createSequentialGroup()
                .addGap(19, 19, 19)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(jLabel1)
                    .addComponent(jLabel4)
                    .addComponent(nmi, javax.swing.GroupLayout.PREFERRED_SIZE, 83, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(averageT, javax.swing.GroupLayout.PREFERRED_SIZE, 83, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addGap(81, 81, 81)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(jPanel2Layout.createSequentialGroup()
                        .addComponent(jLabel5)
                        .addContainerGap(243, Short.MAX_VALUE))
                    .addGroup(jPanel2Layout.createSequentialGroup()
                        .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING, false)
                            .addComponent(thb, javax.swing.GroupLayout.DEFAULT_SIZE, 83, Short.MAX_VALUE)
                            .addComponent(jLabel2, javax.swing.GroupLayout.Alignment.LEADING)
                            .addComponent(kendall, javax.swing.GroupLayout.Alignment.LEADING))
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addComponent(jLabel3)
                            .addComponent(spearman, javax.swing.GroupLayout.PREFERRED_SIZE, 84, javax.swing.GroupLayout.PREFERRED_SIZE))
                        .addGap(15, 15, 15))))
        );
        jPanel2Layout.setVerticalGroup(
            jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(jPanel2Layout.createSequentialGroup()
                .addGap(18, 18, 18)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel1)
                    .addComponent(jLabel2)
                    .addComponent(jLabel3))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(nmi, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(kendall, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(spearman, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addGap(32, 32, 32)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(jLabel4)
                    .addComponent(jLabel5))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(jPanel2Layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                    .addComponent(averageT, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addComponent(thb, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addContainerGap(23, Short.MAX_VALUE))
        );


        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addComponent(runEnGNetButton, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(explanationLabel, javax.swing.GroupLayout.PREFERRED_SIZE, 0, Short.MAX_VALUE)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(loadDatasetButton)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                        .addComponent(genesFileTextField, javax.swing.GroupLayout.DEFAULT_SIZE, 50, Short.MAX_VALUE))
                    .addComponent(jPanel2, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addComponent(explanationLabel, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                		.addComponent(loadDatasetButton)
                		.addComponent(genesFileTextField, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(jPanel2, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addComponent(runEnGNetButton, javax.swing.GroupLayout.PREFERRED_SIZE, 44, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addContainerGap(javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
        );
        
        genesFileTextField.setVisible(false);

        pack();
    }// </editor-fold>//GEN-END:initComponents

    private void runEnGNetButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_runEnGNetButtonActionPerformed
    	setVisible(false);
    	try {
    	double fNMI = (double)this.nmi.getValue(); 
    	double fKendall = (double)this.kendall.getValue();
    	double fSpearman = (double)this.spearman.getValue();
    	double fAverage = (double)this.averageT.getValue();
    	int fThb = (int)this.thb.getValue();
    	String sPathEntrada = this.dataFilePath;
    	File sPath = this.currentFolder;
    	
    	
        taskManager.execute(new org.cytoscape.work.TaskIterator(new Task[] { new ExecuteEnGNetTask(sPath, sPathEntrada, fNMI, fKendall, fSpearman, fAverage, fThb, this) }));
    	}catch(Exception e) {
    		CySwing2.displayPopUpMessage("ERROR:" + e.getMessage());
    	}
    	
    	}//GEN-LAST:event_runEnGNetButtonActionPerformed

    
    private void loadDatasetButtonActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_loadDatasetButtonActionPerformed
    	JFileChooser fileChooser = new JFileChooser();
        fileChooser.setCurrentDirectory(currentFolder);
        int returnValue = fileChooser.showOpenDialog(this);
        if (returnValue != 0) {
            return;
        }
        
        currentFolder = new File(fileChooser.getSelectedFile().getParent());
        dataFilePath= fileChooser.getSelectedFile().getAbsolutePath();
 
        loadDatasetButton.setText("Change dataset file");
        genesFileTextField.setText(dataFilePath);
        genesFileTextField.setVisible(true);
    }//GEN-LAST:event_loadDatasetButtonActionPerformed
    
    

    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JSpinner averageT;
    private javax.swing.JLabel explanationLabel;
    private javax.swing.JButton loadDatasetButton;
    private javax.swing.JLayeredPane datasetsPanel;
    private javax.swing.JLabel jLabel1;
    private javax.swing.JLabel jLabel2;
    private javax.swing.JLabel jLabel3;
    private javax.swing.JLabel jLabel4;
    private javax.swing.JLabel jLabel5;
    private javax.swing.JSpinner kendall;
    private java.awt.Label label1;
    private javax.swing.JSpinner nmi;
    private javax.swing.JButton runEnGNetButton;
    private javax.swing.JSpinner spearman;
    private javax.swing.JSpinner thb;
    private javax.swing.JScrollPane jScrollPane1;
    private javax.swing.JPanel jPanel2;
    private javax.swing.JTextField genesFileTextField;
    // End of variables declaration//GEN-END:variables
}
